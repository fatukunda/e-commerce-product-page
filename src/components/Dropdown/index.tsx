import Menu from "../Menu/Menu";

export interface IOption {
  id: number;
  name: string;
}
interface IFilterProps {
  options: IOption[];
  change: Function;
  selectedItem: IOption;
  close: Function;
  isOpen: boolean;
  toggleOpen: Function;
}

const Dropdown = ({
  options,
  change,
  selectedItem,
  isOpen,
  toggleOpen,
}: IFilterProps): JSX.Element => {
  return (
    <div className="flex flex-col w-48">
      <button
        onClick={() => toggleOpen()}
        className=" flex bg-white items-center justify-between btn-sm w-48 text-xs hover:bg-gray-100 text-tertiary-light font-semibold py-2 px-4  border border-tertiary-light rounded shadow"
        data-testid="select-filter-btn"
      >
        <span>{selectedItem.name}</span>
        {isOpen ? (
          <svg style={{ width: "24px", height: "24px" }} viewBox="0 0 24 24">
            <path
              fill="currentColor"
              d="M7.41,15.41L12,10.83L16.59,15.41L18,14L12,8L6,14L7.41,15.41Z"
            />
          </svg>
        ) : (
          <svg style={{ width: "24px", height: "24px" }} viewBox="0 0 24 24">
            <path
              fill="currentColor"
              d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"
            />
          </svg>
        )}
      </button>
      {isOpen && (
        <Menu classList="bg-white w-48 p-4 text-tertiary-light text-sm font-normal border border-tertiary-light rounded">
          <div className="grid grid-cols-1 divide-y">
            {options.map((option) => (
              <div
                onClick={() => change(option)}
                key={option.id}
                className="border-tertiary-light pb-2 pt-2 hover:bg-gray-100 cursor-pointer"
              >
                {option.name}
              </div>
            ))}
          </div>
        </Menu>
      )}
    </div>
  );
};

export default Dropdown;
