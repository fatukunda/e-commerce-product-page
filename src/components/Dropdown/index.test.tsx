import { render, cleanup, fireEvent, screen} from "@testing-library/react";
import renderer from "react-test-renderer";
import { Provider as ReduxProvider } from "react-redux";

import Dropdown, { IOption } from ".";
import mockStore from "src/components/mocks/store";

const options: IOption[] = [
  {
    id: 1,
    name: "Option 1",
  },
  {
    id: 2,
    name: "Option 2",
  },
  {
    id: 3,
    name: "Option 3",
  },
];

const selectOption = jest.fn()
const toggleDropdown = jest.fn()

const renderDropdown = () =>
  render(
    <ReduxProvider store={mockStore()}>
      <Dropdown
        options={options}
        change={selectOption}
        selectedItem={options[0]}
        isOpen={true}
        toggleOpen={toggleDropdown}
        close={() => jest.fn()}
      />
    </ReduxProvider>
  );

describe("Tests for <Dropdown/>", () => {
  afterEach(cleanup);
  test("Dropdown: Snapshot", () => {
    const component = renderer.create(
      <ReduxProvider store={mockStore()}>
        <Dropdown
          options={options}
          change={() => jest.fn()}
          selectedItem={options[0]}
          isOpen={false}
          toggleOpen={() => jest.fn()}
          close={() => jest.fn()}
        />
      </ReduxProvider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("Renders without crashing", () => {
    renderDropdown();
  });
  it("Toggles dropdown", () => {
    const { getByTitle } = renderDropdown();
    const dropdownButton = getByTitle("select-filter-btn");
    const selectedItem = getByTitle('selected-item')
    expect(selectedItem.textContent).toBe('Option 1')
    fireEvent.click(dropdownButton)
    expect(toggleDropdown).toHaveBeenCalled()
  });
  it('Selects an option when clicked', () => {
    const { getAllByTitle } = renderDropdown();
    const options = getAllByTitle('option')
    fireEvent.click(options[1])
    expect(selectOption).toHaveBeenCalled()
  })
});
