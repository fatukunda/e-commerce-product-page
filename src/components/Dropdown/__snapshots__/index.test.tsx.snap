// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <Dropdown/> Dropdown: Snapshot 1`] = `
<div
  className="flex flex-col w-48"
>
  <button
    className="flex bg-white items-center justify-between 
  btn-sm w-48 text-xs hover:bg-gray-100 text-tertiary-light font-semibold 
  py-2 px-4  border border-tertiary-light rounded shadow"
    onClick={[Function]}
    title="select-filter-btn"
  >
    <span
      title="selected-item"
    >
      Option 1
    </span>
    <svg
      style={
        Object {
          "height": "24px",
          "width": "24px",
        }
      }
      viewBox="0 0 24 24"
    >
      <path
        d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;
